### PR TITLE
small_vector used in Vertex, 24B reduction

### PR DESCRIPTION
**Pull request by Darych**
### Description

Introduce a new small vector datatype (16B) with a small buffer optimization.
Use it for Vertex to have smaller labels collection and smaller edge collections

[master < Task] PR
- [x] Provide the full content or a guide for the final git message
    - **New small vector (16B) and use to save 24B per Vertex**

### Documentation checklist
- [x] Add the documentation label tag
- [x] Add the bug / feature label tag
- [x] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [ ] Write a release note, including added/changed clauses
    - **[Release note text]**
- [ ] Link the documentation PR here
    - **[Documentation PR link]**
- [x] Tag someone from docs team in the comments

### CODE REVIEW

File: src/storage/v2/compact_vector.hpp [hunk 1]
```diff
@@ -315,13 +190,237 @@ struct CompactVector {
...
+      : size_(std::distance(first, last)), capacity_{std::max(size_, kSmallCapacity)} {
+    if (!usingSmallBuffer(capacity_)) {
+      buffer_ = reinterpret_cast<pointer>(std::aligned_alloc(alignof(T), capacity_ * sizeof(T)));
+    }
+    std::ranges::uninitialized_copy(first, last, begin(), end());
+  }
+
+  void push_back(const_reference value) {
+    if (capacity_ == size_) {
+      reserve(capacity_ == 0 ? 1 : 2 * capacity_);
```
Comment (Darych) [inline src/storage/v2/compact_vector.hpp:324]: I think `capacity_` now can't be 0. It's minimum `kSmallCapacity`.

File: src/storage/v2/compact_vector.hpp [hunk 2]
```diff
@@ -315,13 +190,237 @@ struct CompactVector {
...
+    // NOTE 2: either move the buffer or move into the small buffer
+    if (usingSmallBuffer(capacity_)) {
+      std::ranges::uninitialized_move(other, *this);
+    } else {
+      buffer_ = std::exchange(other.buffer_, nullptr);
+    }
+  }
+
+  auto operator=(const CompactVector &other) -> CompactVector & {
+    if (std::addressof(other) != this) {
```
[the remark this answers is not in the document]
Comment (Ignition) [reply src/storage/v2/compact_vector.hpp:216]: Agree but I'll also add `[[unlikely]]`

File: src/storage/v2/compact_vector.hpp (hunk 1 above)
Comment (Ignition) [reply src/storage/v2/compact_vector.hpp:324]: It can be 0 if `8 < sizeof(T)`

**Review by gvolfing — APPROVED** [review]
Really nice!
Just a couple of non-blocking questions, it is probably out of scope but do we plan on templatizing the allocator as well? Not sure how it is playing together with the recent pmr related changes.
Should we add `constexpr` to the member and friend functions?

File: src/storage/v2/vertex.hpp [hunk 3]
```diff
@@ -47,6 +50,7 @@ struct Vertex {
 };
 
 static_assert(alignof(Vertex) >= 8, "The Vertex should be aligned to at least 8!");
+static_assert(sizeof(Vertex) == 8 + 16 + 16 + 16 + 12 + 4 + 8 + 8);
```
[the remark this answers is not in the document]
Comment (Ignition) [reply src/storage/v2/vertex.hpp]: Changed with a better maintenance static_assert 
```cpp
static_assert(sizeof(Vertex) == 88, "If this changes documentation needs changing");
```

**Comment by Ignition** [comment]
@kgolubic tagged for documentation
We need to change https://memgraph.com/docs/fundamentals/storage-memory-usage 

> Each Vertex object has at least 112B + 56B for the Delta object, in total, a minimum of 168B.

112B is now 88B

**Comment by kgolubic** [comment]
Docs PR https://github.com/memgraph/documentation/issues/790

**Pull request merged**